source/server/hot_restart_impl.cc

```diff
@@ -138,319 +137,35 @@ HotRestartImpl::HotRestartImpl(const Options& options)
-  }
-
   // If our parent ever goes away just terminate us so that we don't have to rely on ops/launching
   // logic killing the entire process tree. We should never exist without our parent.
   int rc = prctl(PR_SET_PDEATHSIG, SIGTERM);
   RELEASE_ASSERT(rc != -1, "");
 }
 
-int HotRestartImpl::bindDomainSocket(uint64_t id) {
-  Api::OsSysCalls& os_sys_calls = Api::OsSysCallsSingleton::get();
-  // This actually creates the socket and binds it. We use the socket in datagram mode so we can
-  // easily read single messages.
-  int fd = socket(AF_UNIX, SOCK_DGRAM | SOCK_NONBLOCK, 0);
-  sockaddr_un address = createDomainSocketAddress(id);
-  Api::SysCallIntResult result =
-      os_sys_calls.bind(fd, reinterpret_cast<sockaddr*>(&address), sizeof(address));
-  if (result.rc_ != 0) {
-    throw EnvoyException(
-        fmt::format("unable to bind domain socket with id={} (see --base-id option)", id));
-  }
-
-  return fd;
-}
-
-void HotRestartImpl::initDomainSocketAddress(sockaddr_un* address) {
-  memset(address, 0, sizeof(*address));
-  address->sun_family = AF_UNIX;
-}
-
-sockaddr_un HotRestartImpl::createDomainSocketAddress(uint64_t id) {
-  // Right now we only allow a maximum of 3 concurrent envoy processes to be running. When the third
-  // starts up it will kill the oldest parent.
-  const uint64_t MAX_CONCURRENT_PROCESSES = 3;
-  id = id % MAX_CONCURRENT_PROCESSES;
-
-  // This creates an anonymous domain socket name (where the first byte of the name of \0).
-  sockaddr_un address;
-  initDomainSocketAddress(&address);
-  StringUtil::strlcpy(&address.sun_path[1],
-                      fmt::format("envoy_domain_socket_{}", options_.baseId() + id).c_str(),
-                      sizeof(address.sun_path) - 1);
-  address.sun_path[0] = 0;
-  return address;
-}
-
 void HotRestartImpl::drainParentListeners() {
-  if (options_.restartEpoch() > 0) {
-    // No reply expected.
-    RpcBase rpc(RpcMessageType::DrainListenersRequest);
-    sendMessage(parent_address_, rpc);
-  }
-
+  as_child_.drainParentListeners();
   // At this point we are initialized and a new Envoy can startup if needed.
   shmem_.flags_ &= ~SharedMemory::Flags::INITIALIZING;
 }
 
 int HotRestartImpl::duplicateParentListenSocket(const std::string& address) {
-  if (options_.restartEpoch() == 0 || parent_terminated_) {
-    return -1;
-  }
-
-  RpcGetListenSocketRequest rpc;
-  ASSERT(address.length() < sizeof(rpc.address_));
-  StringUtil::strlcpy(rpc.address_, address.c_str(), sizeof(rpc.address_));
-  sendMessage(parent_address_, rpc);
-  RpcGetListenSocketReply* reply =
-      receiveTypedRpc<RpcGetListenSocketReply, RpcMessageType::GetListenSocketReply>();
-  return reply->fd_;
+  return as_child_.duplicateParentListenSocket(address);
 }
 
-void HotRestartImpl::getParentStats(GetParentStatsInfo& info) {
-  // There exists a race condition during hot restart involving fetching parent stats. It looks like
```

Comment:
This race condition is removed by these changes.

Reply:
Nice!